Regression: Cluster with custom similarity settings fails to start after upgrade

When trying to upgrade an index with custom similarity settings from `7.1.0`, the node fails to start with an exception similar to `
"Caused by: java.lang.IllegalArgumentException: Unknown settings for similarity of type [BM25]: [normalization.h2.c, normalization, after_effect, basic_model]"` . We observed this happening upgrading from `7.1.0` to `7.5.0` and also from `7.1.0` to `7.1.1` . This appears to be the same issue as #25350.

To reproduce,  start up a `7.1.0` cluster, and apply the following settings:
```
PUT /i
{
  "settings": {
    "index": {
      "similarity": {
        "my_similarity": {
          "after_effect": "l",
          "basic_model": "g",
          "normalization": "h2",
          "normalization.h2.c": "3.0",
          "type": "DFR"
        }
      }
    }
  }
}
```

Then attempt to upgrade that cluster.   (Thank you @DaveCTurner for concise repro steps)

--- a/server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java
+++ b/server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java
@@ -22,6 +22,7 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.search.similarities.BM25Similarity;
 import org.apache.lucene.search.similarities.Similarity;
 import org.elasticsearch.Version;
 import org.elasticsearch.common.TriFunction;
@@ -145,7 +146,7 @@ public class MetaDataIndexUpgradeService {
                 @Override
                 public TriFunction<Settings, Version, ScriptService, Similarity> get(Object key) {
                     assert key instanceof String : "key must be a string but was: " + key.getClass();
-                    return SimilarityService.BUILT_IN.get(SimilarityService.DEFAULT_SIMILARITY);
+                    return (settings, version, scriptService) -> new BM25Similarity();
                 }
 
                 // this entrySet impl isn't fully correct but necessary as SimilarityService will iterate

--- a/server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeServiceTests.java
@@ -82,6 +82,18 @@ public class MetaDataIndexUpgradeServiceTests extends ESTestCase {
         assertSame(src, service.upgradeIndexMetaData(src, Version.CURRENT.minimumIndexCompatibilityVersion())); // no double upgrade
     }
 
+    public void testUpgradeCustomSimilarity() {
+        MetaDataIndexUpgradeService service = getMetaDataIndexUpgradeService();
+        IndexMetaData src = newIndexMeta("foo",
+            Settings.builder()
+                .put("index.similarity.my_similarity.type", "DFR")
+                .put("index.similarity.my_similarity.after_effect", "l")
+                .build());
+        assertFalse(service.isUpgraded(src));
+        src = service.upgradeIndexMetaData(src, Version.CURRENT.minimumIndexCompatibilityVersion());
+        assertTrue(service.isUpgraded(src));
+    }
+
     public void testIsUpgraded() {
         MetaDataIndexUpgradeService service = getMetaDataIndexUpgradeService();
         IndexMetaData src = newIndexMeta("foo", Settings.builder().put("index.refresh_interval", "-200").build());

